README.md
# aurelia-gravatar-component

[![Build Status](https://travis-ci.org/SpoonX/aurelia-gravatar-component.svg?branch=master)](https://travis-ci.org/SpoonX/aurelia-gravatar-component)
[![Known Vulnerabilities](https://snyk.io/test/npm/name/badge.svg)](https://snyk.io/test/npm/aurelia-gravatar-component)
[![Gitter](https://img.shields.io/gitter/room/nwjs/nw.js.svg?maxAge=2592000?style=plastic)](https://gitter.im/SpoonX/Dev)

> This library is a plugin for the [Aurelia](http://www.aurelia.io/) platform.

Simple gravatar component for Aurelia. Based on these specs:
https://secure.gravatar.com/site/implement/images/

## Usage

It globalizes a custom element. A very simple example which should get you
going.

```html
<gravatar
  email.bind="vm.email"
  size.bind="200"
  rating="x"
  secure.bind="true">
</gravatar>

```

## Bindable properties

### email
[string='']
Used to determine what gravatar to fetch from gravatar's api.

### size
> [number=200]
The size of a side in pixels of the square gravatar image.

### rating
> [string='g']
`g`   suitable for display on all websites with any audience type.
`pg`  may contain rude gestures, provocatively dressed individuals, the lesser swear words, or mild violence.
`r`   may contain such things as harsh profanity, intense violence, nudity, or hard drug use.
`x`   may contain hardcore sexual imagery or extremely disturbing violence.

### forceDefault
> [boolean=false]
When true always loads the default image instead of the user uploaded one.

### defaultTo
> [string="monsterid"]
The image to fetch when user has not been found. These can be the following
options.  `404` `mm` `identicon` `monsterid` `wavatar` `retro` `blank`.

### secure
> [boolean=false]

Whether to perform request over https

## Installation

### Aureli-Cli

Run `npm i aurelia-gravatar-component --save` from your project root.

aurelia-gravatar-component makes use of `md5`. So, add following to the `build.bundles.dependencies` section of `aurelia-project/aurelia.json`.

```js
"dependencies": [
  // ...
  "md5",
  {
    "name": "aurelia-gravatar-component",
    "path": "../node_modules/aurelia-gravatar-component/dist/amd",
    "main": "aurelia-gravatar-component",
    "resources": [
      "gravatar.html"
    ]
  },
  // ...
],
```

### Jspm

Run `jspm i aurelia-gravatar-component` from your project root.

aurelia-gravatar-component makes use of `md5`. So, add following to the desired `includes` section of `build/bundles.js`, eg:

```js
"aurelia": {
  "includes": [
    //...
    "md5",
    "aurelia-gravatar-component",
    "[aurelia-gravatar-component/**/*.js]",
    "aurelia-gravatar-component/**/*.html!text",
    //...
  ]
}
```

If the installation results in having forks, try resolving them by running:

```sh
jspm inspect --forks
jspm resolve --only registry:package-name@version
```

### Webpack

Run `npm i aurelia-gravatar-component --save` from your project root.

And add `aurelia-gravatar-component` in the `coreBundles.aurelia` section of your `webpack.config.js`.

### Typescript

Npm-based installations pick up the typings automatically. For Jspm-based installations, run `typings i github:spoonx/aurelia-gravatar-component` or add `"aurelia-gravatar-component": "github:spoonx/aurelia-gravatar-component",` to your `typings.json` and run `typings i`.


doc/CHANGELOG.md
### 0.0.3 (2016-06-30)


#### Bug Fixes

* **index:** import used Config class ([580c399a](https://github.com/SpoonX/aurelia-gravatar-component/commit/580c399a87703c1ee7fae7c96b726dbfcfb722da))


### 0.0.2 (2016-06-20)


#### Bug Fixes

* **dependencies:** removed unused and updated ([1468463b](https://github.com/SpoonX/aurelia-gravatar-component/commit/1468463bbf1a5a3352f763501b1f3537f509b17a))
* **project:** code fixes and working tests ([0c6fc63f](https://github.com/SpoonX/aurelia-gravatar-component/commit/0c6fc63fdc62a18eda41b5aa1e6b8eff0a531136))


test/setup.js
// import 'aurelia-polyfills';
// import {initialize} from 'aurelia-pal-browser';
// initialize();


src/config.spec.js
import {GravatarCustomElement} from './gravatar';
import {Config} from './config';

describe('Config', () => {

  let gravatar;

  beforeEach(() => {
    gravatar = new GravatarCustomElement();
  });

  it('has the following defaults', () => {
    expect(gravatar.email).toBe('');
    expect(gravatar.size).toBe(200);
    expect(gravatar.defaultTo).toBe('monsterid');
    expect(gravatar.forceDefault).toBe(false);
    expect(gravatar.secure).toBe(false);
  });

  it('throws an error when default to image is not a valid one', () => {
    gravatar.defaultTo = 'bingo';
    expect(() => {gravatar.url;}).toThrow();
  });

  it('does not throw an error when is a valid default', () => {
    gravatar.defaultTo = 'identicon';
    expect(() => {gravatar.url;}).not.toThrow();
  });

  it('can force default image to load by appending &=y', () => {
      gravatar.forceDefault = true;
      expect(gravatar.url.endsWith('&f=y')).toBe(true);
  });

  it('constructs the url correctly', () => {
    expect(gravatar.url).toBe('http://www.gravatar.com/avatar/d41d8cd98f00b204e9800998ecf8427e/?s=200&d=monsterid&r=g');
  });
});


test/config.spec.js
import {GravatarCustomElement} from 'src/gravatar';
import {Config} from 'src/config';

describe('Config', () => {

  let gravatar;

  beforeEach(() => {
    gravatar = new GravatarCustomElement(new Config());
  });

  it('has the following defaults', () => {
    expect(gravatar.email).toBe('');
    expect(gravatar.size).toBe(200);
    expect(gravatar.defaultTo).toBe('monsterid');
    expect(gravatar.forceDefault).toBe(false);
    expect(gravatar.secure).toBe(false);
  });

  it('throws an error when default to image is not a valid one', () => {
    gravatar.defaultTo = 'bingo';
    expect(() => {gravatar.url;}).toThrow();
  });

  it('does not throw an error when is a valid default', () => {
    gravatar.defaultTo = 'identicon';
    expect(() => {gravatar.url;}).not.toThrow();
  });

  it('can force default image to load by appending &=y', () => {
      gravatar.forceDefault = true;
      expect(gravatar.url.endsWith('&f=y')).toBe(true);
  });

  it('constructs the url correctly', () => {
    expect(gravatar.url).toBe('http://www.gravatar.com/avatar/d41d8cd98f00b204e9800998ecf8427e/?s=200&d=monsterid&r=g');
  });
});


spoonx.js
/************************************************/
/* spoonx-tools configuration                   */
/* @see https://github.com/SpoonX/spoonx-tools  */
/************************************************/

var appRoot = 'src/';

module.exports = {
  path: {
    root: appRoot,

    /* options and their defaults */

    /* js files to ignore
     *
     *  ignore: [],
     */

    /* future use: use TypeScript or Babel for transpiling
     *
     * useTypeScriptForDTS: false,
     */

    /* Imports to append to the import block of the main file. 
     * Add here eg. non-concated local imports in the main file as they will 
     * get removed during the build process (ValueConverters, CustomElements).
     *
     * importsToAdd: ["import {AssociationSelect} from './association-select';"],
     */

    /* js to be transpiled, but not be concated
     * (ValueConverters, CustomElements)
     *
     * jsResources: [appRoot + 'association-select.js'],
     */
    jsResources: [appRoot + 'gravatar.js'],

    /* other resources that need to get copied keeping their path
     * resources: appRoot + '{** / *.css,** / *.html}',
     */
    resources: appRoot + '{**/*.css,**/*.html}',

    /* imports that are only used internally, eg 'extend'. no need to d.ts export them
     *
     * importsToIgnoreForDts: ['extend'],
     */
    importsToIgnoreForDts: ['md5'],

    /* sort when concating
     * sort: true,
     */
    sort: true,

    /* concat js files
     * concat: true,
     */
    concat: true,

    /* default options overwrites for karma
     * karma: {browsers: ['Chrome']}
     */
  }
};


karma.conf.js
// Karma configuration
// Generated on Fri Dec 05 2014 16:49:29 GMT-0500 (EST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jspm', 'jasmine'],

    jspm: {
      // Edit this to your needs
      loadFiles: ['test/**/*.js', 'src/**/*.spec.js'],
      serveFiles: ['src/**/*.js'],
      paths: {
        '*': '*',
        'github:*': 'jspm_packages/github/*',
        'npm:*': 'jspm_packages/npm/*'
      }
    },

    // list of files / patterns to load in the browser
    files: [],

    // list of files to exclude
    exclude: [],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'test/**/*.js': ['babel'],
      'src/**/*.js': ['babel']
    },
    'babelPreprocessor': {
      options: {
        sourceMap: 'inline',
        presets: [ 'es2015-loose', 'stage-1'],
        plugins: [
          'syntax-flow',
          'transform-decorators-legacy',
          'transform-flow-strip-types'
        ]
      }
    },

    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],

    // web server port
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['PhantomJS'],

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
  });
};


src/gravatar.js
import {computedFrom, bindable, inject} from 'aurelia-framework';
import {Config} from './config';
import md5 from 'md5';

@inject(Config)
export class GravatarCustomElement {

  @bindable email        = '';
  @bindable size         = 200;
  @bindable rating       = 'g';
  @bindable forceDefault = false;
  @bindable defaultTo    = 'monsterid';
  @bindable secure       = false;

  constructor(config) {
    Object.assign(this, config.configurations); /* alter the defaults of the view model */
  }

  @computedFrom('defaultTo')
  get defaultGravatar() {
    const defaults = ['404', 'mm', 'identicon', 'monsterid', 'wavatar', 'retro', 'blank'];

    if (defaults.indexOf(this.defaultTo) === -1) {
      throw new Error(`${this.defaultTo} is not a valid gravatar default to image. Choose one of the following: ` + JSON.stringify(defaults));
    }

    return this.defaultTo;
  }

  @computedFrom('size', 'email', 'defaultGravatar', 'forceDefault', 'secure')
  get url() {
    let base = (this.secure ? 'https://secure' : 'http://www') + '.gravatar.com/avatar';
    let hash = md5(this.email);
    let url  = `${base}/${hash}/?s=${this.size}&d=${this.defaultGravatar}&r=${this.rating}`;

    return url + (this.forceDefault ? '&f=y' : '');
  }
}
